cli/2024-06-04-aws-irsa-support.md

```diff
@@ -0,0 +1,376 @@
+
+  @doc("The AWS IRSA credential")
+  IRSA,
+}
+```
+
+```
+model AwsIRSACredentialProperties extends AwsCredentialProperties {
+  @doc("Access Key kind")
+  kind: AWSCredentialKind.IRSA;
+
+  @doc("RoleARN for AWS IRSA identity")
+  @secret
+  roleARN: string;
+
+  @doc("The storage properties")
+  storage: CredentialStorageProperties;
+}
+```
+
+Based on this new model we have to update API calls. 
+
+### CLI Design
+
+* `rad init --full` should be updated to ask the user to choose between AWS access keys and IRSA. If IRSA is chosen, the user should be able to input a AWS IAM Role ARN. As a result of rad init, 
+   1. the helm charts should be updated with service-account token mounted as secret.
+   2. roleARN should be stored as a AWS credential
+
+* `rad install kubernetes --set global.aws.IRSA.enabled=true` should add the neccessary pod spec to mount the service-account token.
+
+* `rad credential register aws --access-key-id <access-key-id> --secret-access-key <secret-access-key>` should be replaced with 
+`rad credential register aws secret --access-key-id <access-key-id> --secret-access-key <secret-access-key>`
+
+* `rad credential register aws irsa --iam-role <roleARN>` should be supported.
+
+### Multi-tenancy
+
+Currently, we support only one credential. However, for future, we might want to support multiple credentials potentially across AWS accounts ( say, one for QA team and other for Dev). The code changes for IRSA should be compatible with these changes in future. Since IRSA setup establishes trust between the cluster - oidc provider and AWS, once we support multiple credentials, we should be able to assume different roleARN based on which environment we are deploying from. We just have to establish right trust entities in all roles.  
+
+### Implementation Details
+
+#### UCP
+
+UCP is responsible for communicating with AWS to deploy AWS resources. 
+UCP should be able to assume the role specified in the credential and use this to communicate with AWS since it has the service-token mounted.
+
+#### Bicep
+
+NA
+
+#### Deployment Engine
+
+NA
+
+#### Core RP  / Recipes RP
+
+Terraform provider requires AWS credentials in order to deploy recipes.
+
+RP should be able to assume the role specified in the credential and use this to communicate with AWS since its pod will have the service-token mounted.
+
+### Error Handling
+
+1. handle mismatch of credential type and arguments supplied.
+2. what if the cluster does not have a oidc provider configured?
+3. what if the user has aws credentials (secrets) configured and wants to switch to IRSA? 
+   1. the user has to re-install radius to disable IRSA. This should print informational messages to unregister IRSA credential and re-register the access key credentials.
+   2. If there is a mismatch (pod spec has IRSA enabled, but credential type is not role-arn or vice-versa, deployment will fail. Enhance error message to suggest checking the right kind of credential is configured)
+
```

Comment:
error - detectable/ not, if we can detect, help user to fix it, otherwise, point a documentation that can help troubleshoot.